### PR TITLE
fix: restore Azure AD bearer token support in api_key auth

### LINKED ISSUE #3282: AzureOpenAI with AAD bearer token passed via api_key works in 2.33.0 but returns 401 in 2.34.0 and after
### Confirm this is an issue with the Python library and not an underlying OpenAI API

- [x] This is an issue with the Python library

### Describe the bug

After upgrading from `openai==2.33.0` to `openai==2.34.0`, requests made with `AzureOpenAI` started returning `401 Unauthorized`.

The same code works with `2.33.0` and fails with `2.34.0`.

In our setup, we obtain an Azure AD access token and pass it into `AzureOpenAI(..., api_key=...)` while using a custom `base_url` that points to an Azure API Management proxy in front of Azure OpenAI.

This behavior appears to have changed between `2.33.0` and `2.34.0`. My suspicion is that `2.34.0` now consistently treats `api_key` as an `api-key` header, whereas our working `2.33.0` behavior effectively allowed this token to work against our endpoint.

(this is my first github issue report, assisted by ai, feedback is welcome ofcourse!)

### Expected behavior

Behavior should remain compatible across the upgrade, or the change should be documented clearly.

Concretely, one of the following would help:
- `>=2.34.0` preserves the prior behavior, or
- the docs / changelog explicitly state that Azure AD bearer tokens must now be passed via `azure_ad_token` or `azure_ad_token_provider`, not `api_key`.

### Actual behavior

With `openai==2.34.0`, the request fails with `401 'Access denied` from our endpoint.
With `openai==2.33.0`, the same request succeeds.

### Additional context

Repository inspection suggests Azure auth behavior changed in `2.34.0`, including changelog items such as:
- `allow explicit Azure auth headers`
- `resolve python auth type checks`
- `preserve selected auth credentials`
- `avoid bearer fallback for admin auth`

From reading the current code, it looks like:
- `azure_ad_token` / `azure_ad_token_provider` send `Authorization: Bearer ...`
- `api_key` sends `api-key: ...`

That would explain why passing an AAD bearer token in `api_key` now returns `401` in our APIM/proxied Azure setup.

### To Reproduce

1. Install `openai==2.33.0`
2. Run the code below: request succeeds
3. Upgrade to `openai==2.34.0`
4. Run the exact same code: request fails with "Error code: 401 - {'error': {'code': '401', 'message': 'Access denied due to invalid subscription key or wrong API endpoint. Make sure to provide a valid key for an active subscription and use a correct regional API endpoint for your resource.'}}"

### Code snippets

```Python
import os
from openai import AzureOpenAI
from azure.identity import ClientSecretCredential

client_id = dbutils.secrets.get(scope="OneLab-SecretScope", key="DataServicePrincipalClientId")
client_secret = dbutils.secrets.get(scope="OneLab-SecretScope", key="DataServicePrincipalClientSecret")

credential = ClientSecretCredential(
    tenant_id="<redacted>",
    client_id=client_id,
    client_secret=client_secret,
)

access_token = credential.get_token("https://cognitiveservices.azure.com/.default")

os.environ["AZURE_OPENAI_TOKEN"] = access_token.token
os.environ["AZURE_OPENAI_VERSION"] = "2024-06-01"

client = AzureOpenAI(
    api_key=os.environ["AZURE_OPENAI_TOKEN"],
    api_version=os.environ["AZURE_OPENAI_VERSION"],
    base_url="<redacted>",
)

completion = client.chat.completions.create(
    model="gpt-5.1",
    messages=[
        {"role": "user", "content": "Hello! Could you solve 3^2?"}
    ]
)

print(completion)
```

### OS

Databricks runtime 18.1 & Spark 4.1.0

### Python version

3.12.3

### Library version

openai v2.34.0

### PR DESCRIPTION
**Pull request by Oxygen56**
Fixes regression in Azure AD bearer token authentication where AAD tokens
passed via `api_key` stopped working after v2.34.0.

## Problem

In v2.34.0, `AzureOpenAI._auth_headers` was added as an override that
unconditionally sends `api_key` as an `api-key` header. This broke
setups where an Azure AD bearer token was passed through the `api_key`
parameter, which was the standard approach before `azure_ad_token` was
introduced.

When using an Azure API Management proxy (or any endpoint expecting
`Authorization: Bearer`), the token being sent as `api-key` instead
of `Authorization: Bearer` results in a 401 error.

## Fix

The `_auth_headers` method in both `AzureOpenAI` and `AsyncAzureOpenAI`
now respects the `bearer_auth` flag from `SecurityOptions` (which
defaults to `True` for all requests). When `bearer_auth` is True and
`azure_ad_token` is not set, `api_key` is sent as an
`Authorization: Bearer` header, restoring the v2.33.0 behavior.

- `azure_ad_token` is still sent as `Authorization: Bearer` (unchanged)
- `api_key` with `bearer_auth=True` is sent as `Authorization: Bearer` (restored)
- `api_key` without `bearer_auth` is sent as `api-key` (unchanged)

Fixes #3282

### CODE REVIEW

**Comment by Oxygen56** [comment]
This update fixes a bug I found while adding test coverage: `_prepare_options()` was unconditionally injecting an `api-key` header regardless of the `bearer_auth` flag, which meant real requests ended up carrying *both* `Authorization: Bearer` and `api-key` headers simultaneously. The Azure endpoint would then use `api-key` (since it takes precedence server-side), effectively nullifying the Bearer fix.

The fix adds a `bearer_auth` check inside `_prepare_options()` (both sync and async) so that when `security["bearer_auth"]` is True, we set `Authorization: Bearer` there and skip `api-key` injection entirely. When `bearer_auth` is False, behavior is unchanged — backward compatible.

Also added 8 unit tests:
- `_auth_headers` behavior with bearer_auth True/False
- azure_ad_token always takes priority
- No-credential edge case
- Default SecurityOptions verification
- Full end-to-end request via respx confirming only Bearer header is sent (no `api-key`)

The last test is what caught the original dual-header issue.

File: src/openai/lib/azure.py [hunk 1]
```diff
@@ -347,11 +347,13 @@ def _get_azure_ad_token(self) -> str | None:
...
 
     @override
-    def _auth_headers(self, security: SecurityOptions) -> dict[str, str]:  # noqa: ARG002
+    def _auth_headers(self, security: SecurityOptions) -> dict[str, str]:
         if self._azure_ad_token is not None:
             return {"Authorization": f"Bearer {self._azure_ad_token}"}
 
         if self.api_key and self.api_key != API_KEY_SENTINEL:
+            if security.get("bearer_auth", False):
+                return {"Authorization": f"Bearer {self.api_key}"}
```
[the remark this answers is not in the document]
Comment (Oxygen56) [reply src/openai/lib/azure.py:356]: Resolved — the latest commit now checks `options.security.get("bearer_auth", False)` in `_prepare_options()` before deciding which header to inject. When `bearer_auth` is True, only `Authorization: Bearer` is set and the `api-key` path is skipped entirely; `_auth_headers()` follows the same logic. Both the sync and async overrides are updated consistently, and `test_full_request_sends_bearer_header_by_default` confirms only the Bearer header is sent end-to-end.

**Comment by marcuswood-oai** [comment]
Thank you for the contribution. We’re closing this PR based on the [authentication decision and migration guidance in #3282](https://github.com/openai/openai-python/issues/3282#issuecomment-5609217980).
